File: test_daniel/controller/ViewController.m
//
//  ViewController.m
//  test_daniel
//
//  Created by daniel on 06/06/20.
//  Copyright © 2020 Daniel. All rights reserved.
//

#import "ViewController.h"
#import "ImageCollectionViewCell.h"
#import "DetailViewController.h"

@interface ViewController ()
@property (strong, nonatomic) IBOutlet UICollectionView *imageCollectionView;
@property (strong, nonatomic) NSArray *images;

@end

@implementation ViewController

- (void)viewWillAppear:(BOOL)animated {
    [self.imageCollectionView registerNib:[UINib nibWithNibName:@"ImageCollectionViewCell" bundle:[NSBundle mainBundle]]
           forCellWithReuseIdentifier:@"ImageCell"];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.images = @[@"Image-0",@"Image-1",@"Image-2",@"Image-3",@"Image-4",@"Image-5",@"Image-6",@"Image-7"];
    [self.imageCollectionView reloadData];
}


- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return self.images.count;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *cellIdentifier = @"ImageCell";
    ImageCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:(cellIdentifier) forIndexPath:indexPath];
    
    cell.image.image = [UIImage imageNamed:([self.images objectAtIndex:indexPath.row])];
    cell.text.text = [self.images objectAtIndex:indexPath.row];
    return cell;
    
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {

   return CGSizeMake((self.view.frame.size.width/2) -7, (self.view.frame.size.width/2));
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    
    [self performSegueWithIdentifier:@"imageToDetail" sender: indexPath];
}

- (void) prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([[segue identifier] isEqualToString:@"imageToDetail"]) {
        
        
        NSIndexPath *index = (NSIndexPath *)sender;
        NSLog(@"%@", index);
        
        DetailViewController* detailViewController = [segue destinationViewController];
        detailViewController.imageName = [self.images objectAtIndex:index.row];
    }
}

@end

File: Experimental/Experimental/Controller/ViewController.swift
//
//  ViewController.swift
//  Experimental
//
//  Created by ogya 1 on 02/05/20.
//  Copyright © 2020 ogya 1. All rights reserved.
//

import UIKit

class ViewController: UIViewController, MovieManagerDelegate {
    
    func didUpdateMovie(movie: MovieModel) {
        item = movie.results
        DispatchQueue.main.async {
            self.collectionView.reloadData()
        }
    }

    @IBOutlet weak var collectionView: UICollectionView!
    
    var collectionViewFlowLayout : UICollectionViewFlowLayout!
    let cellIdentifier = "ItemCollectionViewCell"
    let movieId = "id"
    
    var item : [Result] = []
    
    var apiManager = ApiManager()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        setupCollectionView()
        apiManager.delegate = self
        apiManager.fetchData()
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        setupCollectionViewItemSize()
    }
    
    private func setupCollectionView(){
        collectionView.delegate = self
        collectionView.dataSource = self
        let nib = UINib(nibName: "ItemCollectionViewCell", bundle: nil)
        collectionView.register(nib, forCellWithReuseIdentifier: cellIdentifier)
    }
    
    private func setupCollectionViewItemSize(){
        if collectionViewFlowLayout == nil{
            let lineSpacing : CGFloat = 2
            let interItemSpacing : CGFloat = 1
            
            collectionViewFlowLayout = UICollectionViewFlowLayout()
            collectionViewFlowLayout.sectionInset = UIEdgeInsets.zero
            collectionViewFlowLayout.scrollDirection = .vertical
            collectionViewFlowLayout.minimumLineSpacing = lineSpacing
            collectionViewFlowLayout.minimumInteritemSpacing = interItemSpacing
            
            collectionView.setCollectionViewLayout(collectionViewFlowLayout, animated: true)
        }
    }
}

extension ViewController: UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout
{
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return item.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: cellIdentifier, for: indexPath) as! ItemCollectionViewCell
        
        let url = URL(string: item[indexPath.item].fetchPosterPatch)
        cell.imgMain.load(url: url!)
        cell.txtTitle.text = item[indexPath.item].original_title
        cell.txtDetail.text = item[indexPath.item].release_date
        return cell
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        let item = sender as! Result
        if segue.identifier == movieId{
            if let vc = segue.destination as? DetailViewController{
                vc.id = item.id
            }
        }
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let item = self.item[indexPath.item]
        performSegue(withIdentifier: movieId, sender: item)
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        let flowayout = collectionViewLayout as? UICollectionViewFlowLayout
        let space: CGFloat = (flowayout?.minimumInteritemSpacing ?? 0.0) + (flowayout?.sectionInset.left ?? 0.0) + (flowayout?.sectionInset.right ?? 0.0)
        let size:CGFloat = (collectionView.frame.size.width - space) / 2.0
        return CGSize(width: size, height: 300)
    }
    
}

// Load image dari url
extension UIImageView {
    func load(url: URL) {
        DispatchQueue.global().async { [weak self] in
            if let data = try? Data(contentsOf: url) {
                if let image = UIImage(data: data) {
                    DispatchQueue.main.async {
                        self?.image = image
                    }
                }
            }
        }
    }
}
